spec/02-integration/15-plugins-iterator/02-correctness_spec.lua

```diff
@@ -0,0 +1,143 @@
+  describe("Plugins Iterator - Ensure correctness #" .. strategy, function()
+    local proxy_client, expected_header, must_not_have_headers
+
+    lazy_teardown(function()
+      helpers.stop_kong()
+      helpers.kill_all()
+      assert(conf_loader(nil, {}))
+    end)
+
+    lazy_setup(function()
+      proxy_client = helpers.proxy_client
+      helpers.stop_kong()
+      helpers.kill_all()
+      assert(conf_loader(nil, {}))
+
+      local ef = EntitiesFactory:setup(strategy)
+      ef.bp.plugins:insert(
+        {
+          name = "response-transformer",
+          -- scope to default route
+          route = { id = ef.route_id },
+          config = {
+            add = {
+              headers = { "response-transformed:true" }
+            }
+          }
+        }
+      )
+      ef.bp.plugins:insert(
+        {
+          name = "correlation-id",
+          -- scope to default route
+          route = { id = ef.route_id },
+          config = {
+            header_name = "correlation-id-added"
+          }
+        }
+      )
+
+      for i = 0, 10 do
+        local service = ef.bp.services:insert {
+          path = "/anything/service-" .. i
+        }
+        ef.bp.routes:insert {
+          hosts = { "route.bar." .. i },
+          service = { id = service.id }
+        }
+        ef.bp.plugins:insert(
+          {
+            name = "correlation-id",
+            service = { id = service.id },
+            config = {
+              header_name = "correlation-id-added-service" .. i
+            }
+          }
+        )
+        ef.bp.plugins:insert(
+          {
+            name = "response-transformer",
+            service = { id = service.id },
+            config = {
+              add = {
+                headers = { "response-transformed-" .. i .. ":true" }
+              }
+            }
+          }
+        )
+      end
+
+      local pf = PluginFactory:setup(ef)
+      -- add a plugin scoped to Consumer, Route and Service
+      expected_header = pf:consumer_route_service()
+      must_not_have_headers = {}
+
+      -- scoped to Consumer, Route
+      insert(must_not_have_headers, (pf:consumer_route()))
+      -- assure we don't iterate over #{}
+      assert.is_equal(#must_not_have_headers, 1)
+
+      assert(helpers.start_kong({
+        database   = strategy,
+        nginx_conf = "spec/fixtures/custom_nginx.template",
+      }))
+    end)
+
+    it("ensure no cross-contamination", function()
+      -- meaning that we don't run plugins that are scoped to a specific service
+      for i = 0, 10 do
+        local r = proxy_client():get("/anything/service-" .. i, {
+          headers = {
+            host = "route.bar." .. i,
+            -- authenticate as `alice`
+            apikey = "alice",
+          },
+        })
+        assert.response(r).has.status(200)
+        -- The plugin for service+1 is not executed
+        assert.request(r).has_no_header("correlation-id-added-service" .. i + 1)
```

Comment:
would it make sense to check *all* the other services (to confirm the plugins did not execute) rather than just the ones immediately before and after?